### ISSUE TITLE
learnErrors selfConsist stepX Killed

### ISSUE THREAD

**Issue by piquan64** [benjjneb/dada2#466]
Hi Ben,

**Potentially similar to #336 

R quits when I try to run learnErrors, but only for my reverse reads. I have tried changing "truncLen ="
in:

> out <- filterAndTrim(fnFs1, filtFs1, fnRs1, filtRs1, truncLen = c(290,210), maxN = 0, maxEE = c(2,2), truncQ = 2, rm.phix = TRUE, compress = TRUE, multithread = TRUE)

which sometimes gets me an additional selfConsist step #.

But inevitably, I get the following:

> errR <- learnErrors(filtRs1, multithread = TRUE)
118430557 total bases in 393457 reads from 7 samples will be used for learning the error rates.
Initializing error rates to maximum possible estimate.
selfConsist step 1 .......
   selfConsist step 2Killed

then R quits.

I have tried reinstalling all relevant packages, and recompiled R this morning. I do not know how to proceed. 

I am working with my own data and running everything through my institution's hpc so memory is not an issue.
 
Forward reads converge after 8 rounds.

R version 3.5.0
Rcpp version 0.12.16
RcppParallel version 4.4.0.9
dada2 version 1.8.0
ShortRead version 1.38.0

Any help is greatly appreciated!

**Comment by benjjneb** [comment]
How are you installing the R packages like Rcpp/ShortRead/dada2? Is it all through Bioconductor?

**Comment by piquan64** [comment]
Thanks for the reply!

dada2 was installed by 

source("https://bioconductor.org/biocLite.R")
biocLite("dada2")

However, I got the following error regarding Rcpp:
>/include ../../src/tbb/x86_rtm_rw_mutex.cpp
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s: Assembler messages:
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s:25: Error: no such instruction: `xend'
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s:240: Error: no such instruction: `xbegin .L18'
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s:359: Error: no such instruction: `xabort $255'
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s:521: Error: no such instruction: `xbegin .L39'
>/var/tmp/pbs.1146842.head1.cm.cluster/ccEvr0IY.s:641: Error: no such instruction: `xabort $255'
>make[2]: *** [x86_rtm_rw_mutex.o] Error 1
>make[2]: Leaving directory >`/var/tmp/pbs.1146842.head1.cm.cluster/RtmpWT1cBQ/devtools40d331d7dc3d/RcppParallel/src/tbb/>build/lib_release'
>make[1]: *** [tbb_release] Error 2
>make[1]: Leaving directory >`/var/tmp/pbs.1146842.head1.cm.cluster/RtmpWT1cBQ/devtools40d331d7dc3d/RcppParallel/src/tbb/>src'
>cp: cannot stat `tbb/build/lib_release/libtbb*.*': No such file or directory
>make: *** [tbb] Error 1
>ERROR: compilation failed for package ‘RcppParallel’
>removing ‘/home/u16/youngerman/R/library/RcppParallel’
>Installation failed: Command failed (1)
>/cm/shared/uaapps/R/3.4.4/lib64/R/bin/R' --no-site-file --no-environ  \
>--no-save --no-restore --quiet CMD INSTALL  \
>'/var/tmp/pbs.1146842.head1.cm.cluster/RtmpWT1cBQ/devtools40d367be6bde/benjjneb-dada2->da9419b'  \
>--library='/home/u16/youngerman/R/library' --install-tests 
>ERROR: dependency ‘RcppParallel’ is not available for package ‘dada2’
>removing ‘/home/u16/youngerman/R/library/dada2’
>Installation failed: Command failed (1)
> install.packages("/home/u16/youngerman/work/dada2-1.6",
>repos = NULL,
>type = "source",
>dependencies = c("Depends", "Suggests","Imports"))
>Installing package into ‘/home/u16/youngerman/R/library’
>(as ‘lib’ is unspecified)
>ERROR: dependency ‘RcppParallel’ is not available for package ‘dada2’
>removing ‘/home/u16/youngerman/R/library/dada2’
>Warning message:
>In install.packages("/home/u16/youngerman/work/dada2-1.6", repos = NULL,  :
>installation of package ‘/home/u16/youngerman/work/dada2-1.6’ had non-zero exit status

This was a compatibility problem with the gcc. I was able to get around this issue using mt1022's method on this thread: https://github.com/RcppCore/RcppParallel/issues/29/ (bottom of the thread).

After doing the appropriate commenting out, I was able to download dada2. All steps prior to 
errR <- learnErrors(filtRs1, multithread = TRUE) worked just fine.

I did not actually download ShortRead, so assume in came in with dada2.

**Comment by benjjneb** [comment]
My concern is that the dada2 package and Rcpp/RcppParallel have been built with different compilers, and this is leading to an incompatibility somewhere within the compiled code. Might that be the case?

One workaround to try would be to turn off multithreading in the `errR <- ` line causing the crash.

But looking at these errors the fundamental problem seems to be arising within the RcppParallel and/or the intel TBB library code, which if so is probably beyond my ability to diagnose. I'd try ensuring the same build tools for each component, and the non-multithreading workaround.

**Comment by piquan64** [comment]
hi Ben,

Thanks for the reply.  Coincidentally, the gcc was just updated here, so I recompiled R and RcppParallel and reinstalled dada2. I also tried running errR with multithread set to FALSE. 

Unfortunately, the problem persists.

I am going to put R, dada2 and dependancies into a Singularity container to see if I can run it that way.

**Comment by benjjneb** [comment]
Did you find a workaround for this?

**Comment by piquan64** [comment]
not yet. should have something worked out by the end of the week.

**Comment by piquan64** [comment]
Success. A colleague created a container in Miniconda3 (there were umpteen
compatibility problems trying to accomplish this with a singularity
container...) through which I could run R, dada2 and its dependancies. I
think part of the problem may have arisen from where I was running dada2.
Like a noob I was doing this on the head node of the HPC. I was getting the
same error issue with the container that I detailed in the gitHub thread.

I created a compute node with ample GB and everything works so far.

Huzah.

Thanks for your help (and for the package).
